engine/crates/operation-checks/src/schema/async_graphql.rs

```diff
@@ -98,9 +103,59 @@ impl From<ServiceDocument> for schema::Schema {
     }
 }
 
-fn extract_type_name(ty: &async_graphql_parser::types::BaseType) -> String {
-    match ty {
-        async_graphql_parser::types::BaseType::Named(name) => name.to_string(),
-        async_graphql_parser::types::BaseType::List(inner) => extract_type_name(&inner.base),
+fn extract_type(top_level_ty: &async_graphql_parser::types::Type) -> (String, WrapperTypes) {
+    let mut ty = top_level_ty;
+    let mut small = 0u8;
+
+    // Fast path that should be taken nearly all the time.
+    for i in 0..4 {
+        match &ty.base {
+            async_graphql_parser::types::BaseType::Named(name) => {
+                if !ty.nullable {
+                    small |= (WrapperType::Required as u8) << (i * 2);
```

Comment:
I'd add a comment on the reasoning here and below

Reply:
Just pushed something.